### PR TITLE
A0-1998: Reworked e2e pipeline

### PR DESCRIPTION
**Pull request by Marcin-Radecki**
# Description

Currently, for both pull requests and merge PR the same pipeline is run. This is now actually needed if we look on the assumptions.
1. On PR, only testing binary is used in e2e tests (with different to test multiple catchup, but I think it was an oversight originally made when adding such tests, nevertheless this test is disabled now)
2. On the other hand, when PR is merged, we also do .1 as a precondition to pushing **production** `aleph-node` image to ECR.

This PR decouples 1. and 2. This is possible since we have the following PR checks in place:
             * Require a pull request before merging,
            * Require status checks to pass before merging (e2e tests),
            * Require branches to be up to date before merging.
The same checks will be done on `release-* branches after this PR is merged.

The main benefit in this PR is to decrease by at least 8 minutes main e2e pipeline. This is because is takes around 8 minutes to **link** `aleph-node`, "thanks" to LTO.  This linking is done on one thread only, and can't be parallelized.

NOTE: new workflow .github/workflows/upload_production_binary.yml has not yet been tested, which I still plan to do so.
EDIT: this has been tested

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

### CODE REVIEW

File: .github/workflows/_build-test-node-and-runtime.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      SECRETS_AWS_MAINNET_SECRET_ACCESS_KEY: ${{ secrets.AWS_MAINNET_SECRET_ACCESS_KEY }}
+      SECRETS_AWS_DEVNET_ACCESS_KEY_ID: ${{ secrets.AWS_DEVNET_ACCESS_KEY_ID }}
+      SECRETS_AWS_DEVNET_SECRET_ACCESS_KEY: ${{ secrets.AWS_DEVNET_SECRET_ACCESS_KEY }}
+    steps:
+      - name: Checkout source code
+        uses: actions/checkout@v3
+
+      - name: Get branch name and commit SHA
+        id: get_branch
+        uses: ./.github/actions/get-branch
```
Comment (Kosiare) [inline .github/workflows/_build-test-node-and-runtime.yml:25]: Shouldn't we use freshly refactored version of this action from https://github.com/Cardinal-Cryptography/aleph-node/pull/1012?
It applies not only to this workflow.

Comment (Marcin-Radecki) [reply .github/workflows/_build-test-node-and-runtime.yml:25]: Yes, but the other PR is nor merged to `main` yet, so I cannot use it here. I'll most likely merge that PR and rebase this one

File: .github/workflows/upload_production_binary.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      - name: Push aleph-node Current Image to Public ECR
+        env:
+          CURRENT_IMAGE: aleph-node
+          ECR_PUSH_IMAGE: public.ecr.aws/p6e8q1z1/aleph-node
+        run: |
+          docker tag ${{ env.CURRENT_IMAGE }} \
+            ${{ env.ECR_PUSH_IMAGE }}:${{ steps.vars.outputs.sha_short }}
+          docker tag ${{ env.CURRENT_IMAGE }} ${{ env.ECR_PUSH_IMAGE }}:latest
+          docker push ${{ env.ECR_PUSH_IMAGE }}:${{ steps.vars.outputs.sha_short }}
+          docker push ${{ env.ECR_PUSH_IMAGE }}:latest
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/upload_production_binary.yml:116]: Good catch. So I'd add condition on which `latest` is pushed only on `main`

File: .github/workflows/upload_production_binary.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
+---
+#  This workflow builds production aleph-node and runtime, check runtime determinism,
+#  and uploads both aleph-node runtime to ECR. It is intended to work only on pushes
+#  to main and release branches. Those branches must be set as protected branches with
+#  following restrictions:
+#  * Require a pull request before merging,
+#  * Require status checks to pass before merging (e2e tests),
+#  * Require branches to be up to date before merging.
+name: Upload production aleph-node to ECR
```
[the remark this answers is not in the document]
Comment (Kosiare) [reply .github/workflows/upload_production_binary.yml:9]: Yes, IMO you're right. We should be more clear in the naming of workflows. We shouldn't use any shortcuts like "Uploading X", it should be more descriptive like "Uploading X to Y using Z" (not literally).

For example:
Name of the workflow: Build and upload production aleph-node to ECR
Job: main (if there's only one job ofc)
Steps: Build binary, build docker image, tag docker image, push docker image to ECR

File: .github/workflows/_build-test-node-and-runtime.yml (hunk 1 above)
Comment (Marcin-Radecki) [reply .github/workflows/_build-test-node-and-runtime.yml]: done

File: .github/workflows/upload_production_binary.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/upload_production_binary.yml]: done

File: .github/workflows/_build-production-node-and-runtime.yml [hunk 4]
```diff
@@ -1,20 +1,12 @@
 ---
+#  This workflow builds production version of aleph-node and aleph-runtime, ie with some
+#  features enabled. Those binaries MUST be used in the production.
 name: Build aleph-node and aleph-runtime
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/_build-production-node-and-runtime.yml]: done

File: .github/workflows/_build-test-node-and-runtime.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
+---
+#  This workflow builds test version of aleph-node and aleph-runtime, ie with some
+#  features enabled. Those binaries MUST NOT be used in the production.
+name: Build test aleph-node and test aleph-runtime
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/_build-test-node-and-runtime.yml]: done

File: .github/workflows/deploy-feature-envs.yaml [hunk 6]
```diff
@@ -32,8 +32,8 @@ jobs:
     if: >
       (github.event.label.name == '[AZERO] DEPLOY-FEATURE-ENV') ||
       (github.event.label.name == '[AZERO] DEPLOY-HOT-FEATURE-ENV')
-    name: Build runtime and aleph-node binary artefacts
-    uses: ./.github/workflows/build-node-and-runtime.yml
+    name: Build production runtime and aleph-node binary artefacts
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/deploy-feature-envs.yaml]: done

File: .github/workflows/upload_production_binary.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/upload_production_binary.yml]: done

**Comment by Marcin-Radecki** [comment]
marcin@marcin-Latitude-3520 ~/g/aleph-node (build-production-binary-only-on-merge)> docker run --rm --name tmp-validatr -v ~/git/aleph-node/.github:/gh public.ecr.aws/p6e8q1z1/github-workflows-validator:0.2.0 validate -p /gh | grep "^E"
EW803: workflow contracts-deploy.yml job build_and_deploy_contracts step 7              Call to non-existing local action './.github/actions/restore-cache'
EW803: workflow contracts-deploy.yml job build_and_deploy_contracts step 21             Call to non-existing local action './.github/actions/post-cache'
EW201: workflow deploy-mainnet.yml                                                      Called variable 'S3BUCKET_FILE' is invalid
marcin@marcin-Latitude-3520 ~/g/aleph-node (build-production-binary-only-on-merge)>

**Review by mikogs — APPROVED** [review]

File: .github/workflows/upload-production-binary.yml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+      - name: Push aleph-node image to ECR
+        env:
+          CURRENT_IMAGE: aleph-node
+          ECR_PUSH_IMAGE: public.ecr.aws/p6e8q1z1/aleph-node
+        run: |
+          docker tag ${{ env.CURRENT_IMAGE }} \
+            ${{ env.ECR_PUSH_IMAGE }}:${{ steps.get-ref-properties.outputs.sha }}
+          docker push ${{ env.ECR_PUSH_IMAGE }}:${{ steps.get-ref-properties.outputs.sha }}
+          if [[ ${{ steps.get-ref-properties.outputs.branch }} -eq "main" ]]; then
```
Comment (mikogs) [inline .github/workflows/upload-production-binary.yml:69]: Can we put that `${{ steps.get-ref-properties.outputs.branch }}` in double quote please?  If it's empty, the script will fail with syntax error

Comment (Marcin-Radecki) [reply .github/workflows/upload-production-binary.yml]: done

File: .github/workflows/push-commit-to-pull-request.yml [hunk 8]
```diff
@@ -1,180 +1,53 @@
...
-name: e2e-tests-main-devnet
+#  This workflow is intended to run only for PRs on main and release branches,
+#  and does following
+#  * it builds test an aleph-node and runtime, wraps it in Docker,
+#  * in parallel, it builds production node and runtime, and check runtime determinism
+#  * it builds an e2e test client, wraps it in Docker,
+#  * in parallel, it starts an aleph-node consensus and run an e2e test on it
+#  * also in parallel, it checks clippy, unit tests and compiles excluded packages
+#  The whole workflow passes only if all e2e tests pass.
+name: PR
```
Comment (Kosiare) [inline .github/workflows/push-commit-to-pull-request.yml:10]: I think we could name it more descriptive than just "PR".

Comment (Marcin-Radecki) [reply .github/workflows/push-commit-to-pull-request.yml]: Please see now, concept changed a bit

File: .github/workflows/run-e2e-tests-and-push-node-to-ecr.yml [hunk 9]
```diff
@@ -728,110 +553,71 @@ jobs:
...
         env:
           # Relative to local-tests/ directory
-          ALEPH_NODE_BINARY: aleph-release-node/aleph-node
+          ALEPH_NODE_BINARY: aleph-test-node/aleph-node
         run: ./.github/scripts/test_multiple_restarts.sh ${{ matrix.pruning }}
 
-  check-runtime-change:
-    name: Inspect whether runtime version has been changed (compared with main)
+  push-node-image-to-ecr:
+    name: Push production node docker image to the ECR
```
Comment (mikogs) [inline .github/workflows/run-e2e-tests-and-push-node-to-ecr.yml:560]: Let's change it to `Push node docker image to ECR`.  That "production" word might suggest that we actually push something to production not that we push production-version of the node image.

**Review by Kosiare — APPROVED** [review]

File: .github/workflows/on-release-branch-commit.yml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ---
-name: Release branch workflows check
-description: Checks if deployment workflows on release and main branches are identical
+#  Checks if deployment workflows on release and main branches are identical
```
Comment (fbielejec) [inline .github/workflows/on-release-branch-commit.yml:2]: can't have a `description` ?

File: .github/workflows/on-main-branch-commit.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Checkout source code
+        uses: actions/checkout@v2
+
+      - name: Run e2e test
+        uses: ./.github/actions/run-e2e-test
+        with:
+          deploy-button: true
+          clean-button: true
+          test-case: button
+        timeout-minutes: 20
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply .github/workflows/on-main-branch-commit.yml]: I've set 15

**Review by Kosiare — APPROVED** [review]

**Review by fbielejec — APPROVED** [review]

**Pull request merged**